File: src/test/java/io/github/freya022/botcommands/test/commands/slash/SlashButton.java
package io.github.freya022.botcommands.test.commands.slash;

import io.github.freya022.botcommands.api.commands.application.ApplicationCommand;
import io.github.freya022.botcommands.api.commands.application.slash.GuildSlashEvent;
import io.github.freya022.botcommands.api.commands.application.slash.annotations.JDASlashCommand;
import io.github.freya022.botcommands.api.commands.application.slash.annotations.SlashOption;
import io.github.freya022.botcommands.api.components.Components;
import io.github.freya022.botcommands.api.components.annotations.JDAButtonListener;
import io.github.freya022.botcommands.api.components.event.ButtonEvent;
import io.github.freya022.botcommands.api.core.service.annotations.Dependencies;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.interactions.components.buttons.ButtonStyle;

import java.util.concurrent.TimeUnit;

@Dependencies(Components.class)
public class SlashButton extends ApplicationCommand {
	private static final String PERSISTENT_BUTTON_LISTENER_NAME = "leBouton";

	@JDASlashCommand(name = "button")
	public void onSlashButton(GuildSlashEvent event, Components components) {
		event.reply("Buttons")
				.addActionRow(
						components.persistentButton(ButtonStyle.PRIMARY, "Persistent button (1 minute timeout)", builder -> {
							builder.setOneUse(true);
							builder.bindTo(PERSISTENT_BUTTON_LISTENER_NAME, System.currentTimeMillis());
							builder.timeout(1, TimeUnit.MINUTES);
						}),
						components.ephemeralButton(ButtonStyle.PRIMARY, "Ephemeral button (1 second timeout)", builder -> {
							builder.bindTo(btnEvt -> btnEvt.deferEdit().queue());
							builder.timeout(1, TimeUnit.SECONDS, () -> event.getHook().editOriginal("Ephemeral expired :/").queue());
						})
				)
				.setEphemeral(true)
				.queue();
	}

	@JDAButtonListener(name = PERSISTENT_BUTTON_LISTENER_NAME)
	public void onPersistentButtonClicked(ButtonEvent event, @SlashOption long timeCreated, JDA jda) {
		event.replyFormat("Button created on %s and I am %s", timeCreated, jda.getSelfUser().getAsTag())
				.setEphemeral(true)
				.queue();
	}
}

File: src/examples/kotlin/io/github/freya022/bot/commands/slash/SlashButton.kt
package io.github.freya022.bot.commands.slash

import dev.minn.jda.ktx.coroutines.await
import dev.minn.jda.ktx.interactions.components.row
import dev.minn.jda.ktx.messages.into
import io.github.freya022.botcommands.api.commands.annotations.Command
import io.github.freya022.botcommands.api.commands.application.ApplicationCommand
import io.github.freya022.botcommands.api.commands.application.CommandScope
import io.github.freya022.botcommands.api.commands.application.slash.GlobalSlashEvent
import io.github.freya022.botcommands.api.commands.application.slash.annotations.JDASlashCommand
import io.github.freya022.botcommands.api.commands.application.slash.annotations.TopLevelSlashCommandData
import io.github.freya022.botcommands.api.components.Button
import io.github.freya022.botcommands.api.components.Buttons
import io.github.freya022.botcommands.api.components.annotations.JDAButtonListener
import io.github.freya022.botcommands.api.components.builder.bindTo
import io.github.freya022.botcommands.api.components.event.ButtonEvent
import net.dv8tion.jda.api.utils.TimeFormat
import java.time.Instant
import kotlin.time.Duration.Companion.seconds
import kotlin.time.toJavaDuration

@Command
class SlashButton(private val buttons: Buttons) : ApplicationCommand() {
    @TopLevelSlashCommandData(scope = CommandScope.GLOBAL)
    @JDASlashCommand(name = "button", description = "Try out the new buttons!")
    suspend fun onSlashButton(event: GlobalSlashEvent) {
        val components: MutableList<Button> = arrayListOf()
        components += buttons.primary("Click me under 5 seconds").ephemeral {
            timeout(5.seconds) {
                event.hook.editOriginalComponents(components.map(Button::asDisabled).row()).queue()
            }
            bindTo { buttonEvent ->
                buttonEvent.editButton(buttonEvent.button.asDisabled()).await() // Coroutines!
            }
        }

        components += buttons.secondary("Click me anytime").persistent {
            bindTo(::onPersistentButtonClick)
        }

        event.replyComponents(components.into())
            .setContent(TimeFormat.RELATIVE.format(Instant.now() + 5.seconds.toJavaDuration()))
            .setEphemeral(true)
            .queue()
    }

    @JDAButtonListener("SlashButton: persistentButton") //ClassName: theButtonPurpose
    suspend fun onPersistentButtonClick(event: ButtonEvent) {
        event.editButton(event.button.asDisabled()).await()
    }
}